src/server/oauth2/OAuth2Client.ts

```diff
@@ -207,4 +219,118 @@ export class OAuth2Client {
             };
         }
     }
+
+    public async refreshUserAccessToken({
+        user,
+        persis,
+    }: {
+        user: IUser,
+        persis: IPersistence,
+    }) {
+        try {
+            const {
+                config: { clientId, clientSecret, refreshTokenUri },
+            } = this;
+            const siteUrl = await this.app
+                    .getAccessors()
+                    .environmentReader.getServerSettings()
+                    .getValueById('Site_Url');
+            const redirectUri = this.app
+                    .getAccessors()
+                    .providedApiEndpoints[0].computedPath.substring(1);
+            const url = new URL(refreshTokenUri);
+
+            const tokensInfo = await this.getAccessTokenForUser(user);
+
+            for (const tokenInfo of tokensInfo) {
+                if (tokenInfo.refreshToken) {
+                    url.searchParams.set('client_id', clientId);
+                    url.searchParams.set('redirect_uri', `${siteUrl}${redirectUri}`);
+                    url.searchParams.set('refresh_token', tokenInfo.refreshToken);
+                    url.searchParams.set('client_secret', clientSecret);
+                    url.searchParams.set('grant_type', IGrantType.RefreshToken);
+
+                    const { content } = await this.app.getAccessors().http.post(url.href);
+
+                    const { access_token, expires_in, refresh_token } = JSON.parse(
+                        content as string,
+                    );
+
+                    if (access_token) {
+                        await this.removeToken({ userId: user.id, persis }),
+                        await this.saveToken({
+                            token: access_token,
+                            expiresAt: expires_in,
+                            refreshToken: refresh_token,
+                            userId: user.id,
+                            persis,
+                        });
+                    }
+                }
+            }
+        } catch (error) {
+            this.app.getLogger().error(error);
+        }
+
+    }
+
+    public async revokeUserAccessToken({ user, persis }: { user: IUser, persis: IPersistence}) {
+        const tokensInfo = await this.getAccessTokenForUser(user);
+
+        const url = new URL(this.config.revokeTokenURI);
+
+        for (const tokenInfo of tokensInfo) {
+            url.searchParams.set('token', tokenInfo.token);
+
+            await this.app.getAccessors().http.post(url.href);
+        }
+
+        this.removeToken({ userId: user.id, persis });
```

Comment:
When calling functions/methods that return promises, we should either `await` or `return` _always_ - here, for instance, even if a method calling `revokeUserAccessToken` uses an `await` **it will end up not waiting until the `removeToken` method finishes running to return the control**, because this call to `removeToken` simply creates a Promise but doesn't act on it